### PR TITLE
Update waitUntilCondition to use a watcher instead of polling

### PR DESCRIPTION
**Pull request by bbeaudreault**

### CODE REVIEW

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java [hunk 1]
```diff
@@ -50,27 +69,10 @@
...
-import java.util.Map;
-import java.util.Objects;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.TimeUnit;
-import java.util.function.Predicate;
 import okhttp3.HttpUrl;
 import okhttp3.Request;
 
-public class BaseOperation<T, L extends KubernetesResourceList, D extends Doneable<T>, R extends Resource<T, D>>
+public class BaseOperation<T extends HasMetadata, L extends KubernetesResourceList, D extends Doneable<T>, R extends Resource<T, D>>
```
Comment (zklapow) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java:75]: is this type constraint backwards compatible?

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/WaitForConditionWatcher.java [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+          future.complete(resource);
+          break;
+        }
+      case DELETED:
+        if (condition.test(null)) {
+          future.complete(null);
+          break;
+        }
+      case ERROR:
+        log.debug("ERROR event for " + resource.getMetadata().getName());
```
Comment (zklapow) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/WaitForConditionWatcher.java:50]: I wonder if we should complete here? its not even clear what this event would represent (is it an error on the resource? an error in the watcher stream?) but seems like we could get stuck here

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java (hunk 1 above)
Comment (bbeaudreault) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java:75]: I think it is, but I figured I'd let upstream comment on that. They haven't been super concerned about compatibility.

Comment (zklapow) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java:75]: 👍

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/WaitForConditionWatcher.java (hunk 2 above)
Comment (bbeaudreault) [reply kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/WaitForConditionWatcher.java:50]: I was having a hard time figuring out what could come through on the ERROR. I noticed some actual error handling which delegated through onClose, so figured these were transient errors. However, lacking docs, I finally found how the native go client handles this. Will replicate

**Comment by bbeaudreault** [comment]
@zklapow this should be ready for one more look.

File: kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java [hunk 3]
```diff
@@ -1068,61 +1084,51 @@ public Boolean isReady() {
...
+    throws InterruptedException {
 
-    T item = get();
+    T item = fromServer().get();
     if (condition.test(item)) {
       return item;
     }
 
-    throw new IllegalArgumentException(type.getSimpleName() + " with name:[" + name + "] in namespace:[" + namespace + "] not found!");
+    long end = System.currentTimeMillis() + timeoutMillis;
```
Comment (zklapow) [inline kubernetes-client/src/main/java/io/fabric8/kubernetes/client/dsl/base/BaseOperation.java:1106]: technically we should use `nanoTime` for this, `currenTimeMillis` is not monotonic and so it doesn't really work for timing operations like this.

**Pull request merged**